### PR TITLE
Feature style panel flex child section

### PR DESCRIPTION
**Pull request by ghost**
Branching from https://github.com/webstudio-is/webstudio-designer/pull/268
Depends on(Don't merge before): https://github.com/webstudio-is/webstudio-designer/pull/313

- add/update toggle-control.
- add floating panel.

Figma of what's to be added/update for reference.

Screenshot 2022-09-25 at 6 26 10 AM

Screenshot 2022-09-25 at 6 26 02 AM

## Before requesting a review

- [x] if the PR is WIP - use draft mode
- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")
- [ ] what kind of review is needed?
  - [ ] conceptual
  - [ ] detailed
  - [ ] with testing

## Test cases

- [ ] step by step interaction description and what is expected to happen

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-designer/blob/main/apps/designer/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `designer/env-check.js` if mandatory

### CODE REVIEW

File: packages/design-system/src/components/toggle-group.tsx [hunk 1]
```diff
@@ -30,11 +31,12 @@ export const Item = styled(ToggleGroupPrimitive.Item, {
...
   "&:hover": { backgroundColor: "$slateA3" },
+  "&:focus": { boxShadow: "0 0 0 $spacing$2 $colors$blue10", zIndex: 1 },
   "&[data-state=on]": {
-    backgroundColor: "$slateA5",
-  },
-  "&:focus": {
-    position: "relative",
-    boxShadow: `0 0 0 2px black`,
+    backgroundColor: "$slate5",
+    "&[data-breakpoint=true]": {
```
Comment (kof) [inline packages/design-system/src/components/toggle-group.tsx:37]: I feel like we talked about it several times by now.  A design system component doesn't know the designer specific stuff like breakpoint. It's business logic related

File: apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { Flex, ToggleGroup, Tooltip } from "@webstudio-is/design-system";
+import { StyleProperty } from "@webstudio-is/react-sdk";
+import { useIsFromCurrentBreakpoint } from "../../shared/use-is-from-current-breakpoint";
+
+export type ToggleGroupControlProps = {
+  property: StyleProperty | StyleProperty[];
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx:6]: is there a concrete use case of passing an array of properties? If not we should simplify and use single property.

File: apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+export const ToggleGroupControl = ({
+  property,
+  value = "",
+  items = [],
+  onValueChange,
+}: ToggleGroupControlProps) => {
+  const isCurrentBreakpoint = useIsFromCurrentBreakpoint(property);
+  return (
+    <ToggleGroup.Root
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx:20]: Lets export the Toggle group in the ds as `export const ToggleGroupRoot` ... we have it currently inconsistent, but I think this is better ... alternatively the way you did it here and then renaming the rest

File: apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  currentStyle,
+  sectionStyle,
+}: RenderCategoryProps) => {
+  const setAlignSelf = setProperty("alignSelf");
+  const setOrder = setProperty("order");
+  const setSizing = createBatchUpdate();
+  const getSizingValue = () => {
+    switch (
+      String(currentStyle.flexGrow?.value) +
+      String(currentStyle.flexShrink?.value)
```
[the remark this answers is not in the document]
Comment (kof) [reply apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:33]: also this function can be outside of cmponent

File: apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+              child: <FlexSizingShrinkIcon />,
+              label: "Shrink if needed",
+              value: "shrink",
+            },
+            {
+              child: (
+                <ValuePickerPopover
+                  title="Sizing"
+                  content={
+                    <Grid
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:152]: a separate component would be much easier to read

File: apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+                  </Flex>
+                </ValuePickerPopover>
+              ),
+              label: "More sizing options",
+              value: "",
+            },
+          ]}
+        />
+      </Grid>
+      <Grid css={{ gridTemplateColumns: "4fr auto" }}>
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:204]: same thing here, would it make sense for it to be separate component? (same file)

File: apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+              case "0":
+              case "1":
+              case "-1": {
+                setOrder(value);
+                break;
+              }
+            }
+          }}
+          value={String(currentStyle.order?.value)}
+          items={[
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:222]: same thing here, would it make sense for it to be separate component? (same file)

Comment (kof) [reply apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:222]: rather a function

File: apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+              label: "Make first",
+              value: "1",
+            },
+            {
+              child: <OrderLastIcon />,
+              label: "Make last",
+              value: "-1",
+            },
+            {
+              child: (
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:239]: this one could be a component

File: packages/icons/src/align-self-baseline.tsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import * as React from "react";
```
Comment (kof) [inline packages/icons/src/align-self-baseline.tsx:1]: we started generating icon components out of SVG, check out the package scripts

File: packages/icons/src/order-first.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+          width="4"
+          height="8"
+          x="10.9"
+          y="4"
+          stroke={color}
+          strokeWidth="2.2"
+          mask="url(#a)"
+          rx="1"
+        />
+        <mask id="b" fill="#fff">
```
Comment (kof) [inline packages/icons/src/order-first.tsx:29]: is svg becoming part of the global document when rendered directly as a component? I wonder if that id is only visible inside the SVG itself or it just became a global id

**Review by kof — CHANGES_REQUESTED** [review]
first round of review done

File: packages/icons/src/order-first.tsx (hunk 10 above)
Comment (kof) [reply packages/icons/src/order-first.tsx:29]: I understand its from figma, my question is if that's ok or is it a document-global id

File: packages/icons/src/align-self-baseline.tsx (hunk 9 above)
Comment (kof) [reply packages/icons/src/align-self-baseline.tsx:1]: cc @TrySound

Comment (TrySound) [reply packages/icons/src/align-self-baseline.tsx:1]: Props are spread and ref is added. See generated icons. Fill on the root should be enough. I dont see the issue.

File: packages/icons/src/order-first.tsx (hunk 10 above)
Comment (TrySound) [reply packages/icons/src/order-first.tsx:29]: We can add prefixes for ids in svg with svgo plugin

File: packages/icons/src/order-last.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+          width="4"
+          height="8"
+          x="10.9"
+          y="4"
+          stroke={color}
+          strokeWidth="2.2"
+          mask="url(#a)"
+          rx="1"
+        />
+        <mask id="b" fill="#fff">
```
Comment (kof) [inline packages/icons/src/order-last.tsx:30]: this one seems like it's doing something, so we can't just remove it, we need the plugin

File: packages/design-system/src/components/toggle-group.tsx [hunk 12]
```diff
@@ -30,11 +31,6 @@ export const Item = styled(ToggleGroupPrimitive.Item, {
...
   "&:last-child": { borderTopRightRadius: 4, borderBottomRightRadius: 4 },
   "&:hover": { backgroundColor: "$slateA3" },
-  "&[data-state=on]": {
-    backgroundColor: "$slateA5",
-  },
-  "&:focus": {
-    position: "relative",
-    boxShadow: `0 0 0 2px black`,
-  },
+  "&:focus": { boxShadow: "0 0 0 $spacing$2 $colors$blue10", zIndex: 1 },
```
Comment (kof) [inline packages/design-system/src/components/toggle-group.tsx:34]: was there again an issue with stacking context or are you fixing it preemptively? I wonder what we should standardized on, `1` feels random.

Maybe the issue was again somewhere else?

File: apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+      value={value}
+      onValueChange={onValueChange}
+      css={{ width: "fit-content" }}
+    >
+      {items.map(({ child, label, value }, index) => {
+        return (
+          <ToggleGroupControlItem
+            key={index}
+            value={value}
+            state={isCurrentBreakpoint ? "set" : undefined}
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx:37]: "set" is also application level thing, sorry, you either need to add that state to a new styled component like a last time, or it's something that we need everywhere and its not "on" state it already has, then you have to explain what use case it has somewhere else

Comment (kof) [reply apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx:37]: ohhh its local component now, ignore  that comment above

File: apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+      })}
+    </ToggleGroupRoot>
+  );
+};
+
+const ToggleGroupControlItem = styled(ToggleGroupItem, {
+  variants: {
+    state: {
+      set: {
+        "&[data-state=on]": {
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx:53]: why do you need to override the on state this way? is that fixing a specificity issue?

File: apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    </ValuePickerPopover>
+  );
+};
+
+const getSizingValue = (flexGrow: string, flexShrink: string) => {
+  if (flexGrow === "0" && flexShrink === "0") return "none";
+  if (flexGrow === "1" && flexShrink === "0") return "grow";
+  if (flexGrow === "0" && flexShrink === "1") return "shrink";
+  return "";
+};
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/sections/flex-child/flex-child.tsx:277]: this file looks now 1000x more readable now

File: apps/designer/app/designer/features/style-panel/shared/property-name.tsx [hunk 16]
```diff
@@ -28,7 +28,6 @@ export const PropertyName = ({ property, label, css }: PropertyProps) => {
                   color: "$hiContrast",
                 }),
           }}
-          htmlFor={property}
```
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/property-name.tsx:31]: that was for clicking on property to result in focusing the field, why removing?

**Comment by kof** [comment]
Please link the new issues that for SVG stuff

1. id auto generation
2. generating

File: packages/design-system/src/components/toggle-group.tsx (hunk 12 above)
Comment (kof) [reply packages/design-system/src/components/toggle-group.tsx:34]: please add that as a comment, super non-obvious

File: apps/designer/app/designer/features/style-panel/shared/property-name.tsx (hunk 16 above)
Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/property-name.tsx:31]: it didn't work? then we need to fix it?

**Review by kof — CHANGES_REQUESTED** [review]
almost there

File: apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx (hunk 14 above)
Comment (kof) [reply apps/designer/app/designer/features/style-panel/controls/toggle/toggle-control.tsx:53]: damn, this isn't great from stitches side ... what options do we have?

File: packages/design-system/src/components/popover.tsx [hunk 17]
```diff
@@ -77,10 +77,14 @@ export const PopoverHeader = ({ title }: PopoverHeaderProps) => {
         <PopoverClose asChild>
           <IconButton
             size="2"
-            css={{ marginRight: "$spacing$5" }}
+            css={{
+              marginRight: "$spacing$5",
+              "&:not(:hover):focus": { background: "none" },
```
Comment (kof) [inline packages/design-system/src/components/popover.tsx:82]: seems like you were trying to achieve this on focus without hover

Screenshot 2022-11-15 at 16 45 51


But in figma it is specified like this:


Screenshot 2022-11-15 at 16 46 45

Either we need to change it in figma by @taylornowotny and change it on IconButton directly or leave it as-is.

Adhoc changing it in one particular popup is wrong.

**Review by kof — APPROVED** [review]
🎉

**Pull request merged**